### PR TITLE
Detect line endings and use them during code generation

### LINKED ISSUE #1477: Fixing I001 turns file with CRLF line endings into mixed line endings
When fixing `I001` import related lint errors the fix introduces `\n` line endings even if the file originally was `\r\n` only. (On windows  `\r\n` line endings are widely used.)

To reproduce:
```bash
$ echo -e -n '"""A file with CRLF line endings."""\r\nimport sys\r\nimport os\r\n' > crlf.py

$ python -c "from pathlib import Path;print(Path('crlf.py').read_bytes())"
b'"""A file with CRLF line endings."""\r\nimport sys\r\nimport os\r\n'

$ ruff crlf.py --select I001 --fix
Found 1 error(s) (1 fixed, 0 remaining).

$ python -c "from pathlib import Path;print(Path('crlf.py').read_bytes())"
b'"""A file with CRLF line endings."""\r\nimport os\nimport sys\n'
```

The last output contains mixed `\r\n` and `\n` line endings. 
It results e.g. in `pylint` subsequently complaining about [mixed-line-endings / C0327](https://pylint.pycqa.org/en/latest/user_guide/messages/convention/mixed-line-endings.html).

If the original file consistently uses a single line ending convention it would be nice if the fix would adhere to the same convention.

### PR DESCRIPTION
**Pull request by squiddy**
Closes #1477

### CODE REVIEW

**Comment by squiddy** [comment]
(Clippy issue is from main)

**Comment by charliermarsh** [comment]
Nice, this looks quite good!

File: src/source_code_style.rs [hunk 1]
```diff
@@ -30,12 +31,18 @@ impl<'a> SourceCodeStyleDetector<'a> {
             .get_or_init(|| detect_quote(self.contents, self.locator).unwrap_or_default())
     }
 
+    pub fn line_ending(&'a self) -> &'a LineEnding {
+        self.line_ending
+            .get_or_init(|| detect_line_ending(self.contents).unwrap_or_default())
+    }
```
Comment (charliermarsh) [inline src/source_code_style.rs:37]: Thank you for reusing this abstraction!

**Comment by rhkleijn** [comment]
Thanks!

**Comment by charliermarsh** [comment]
Looking back at this: do we also need to change SourceCodeGenerator? That writes newlines too.

**Comment by squiddy** [comment]
> Looking back at this: do we also need to change SourceCodeGenerator? That writes newlines too.

Whoops. I specifically searched for line endings in the code, not sure why the generator slipped through. I can do that follow-up if you want.

**Comment by charliermarsh** [comment]
That’d be great! No rush.

**Pull request merged**